### PR TITLE
Allow to have a better version prediction for functions

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
And as promised, here it is! However the code isn't really nice so if you have any ideas of improvement, I'll gladly update.

### CODE REVIEW

**Comment by EPashkin** [comment]
PR successfully found version increase `gdk_pixbuf_loader_write_bytes` and many increases in gio-sys.
All of these now same as object where it belong, Ex. g_subprocess_communicate_async, g_test_dbus_add_service_dir.
Not added any new versions to normal crates and detect only `g_notification_set_priority` from version overrides in your `Gir.toml` files.

I have worry how this works with crates with nonuniform versions,
maybe this update need be configured and disabled by default.

File: src/update_version.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            (Some(cur_ver), Some(lib_ver)) => cur_ver > lib_ver,
+            (Some(_), None) => true,
+            _ => false,
+        } {
+            function.version = current_version;
+        }
+    }
+}
+
+pub fn check_function_real_version(library: &mut Library) {
```
Comment (EPashkin) [inline src/update_version.rs]: Please place this public function first

File: src/update_version.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                _ => None,
+            } {
+                current_version = Some(new_version);
+            }
+        }
+        if match (current_version, function.version) {
+            (Some(cur_ver), Some(lib_ver)) => cur_ver > lib_ver,
+            (Some(_), None) => true,
+            _ => false,
+        } {
```
Comment (EPashkin) [inline src/update_version.rs]: Maybe better move match to variable

File: src/update_version.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        for parameter in &function.parameters {
+            if let Some(new_version) = match (current_version,
+                                              match *unsafe { (*lib).type_(parameter.typ) } {
+                library::Type::Class(ref c) => c.version,
+                library::Type::Enumeration(ref c) => c.version,
+                library::Type::Bitfield(ref c) => c.version,
+                library::Type::Record(ref c) => c.version,
+                library::Type::Interface(ref c) => c.version,
+                _ => None,
+            }) {
```
Comment (EPashkin) [inline src/update_version.rs]: This also look tough to understand and better splitted

File: src/update_version.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
+use library::{self, Function, Type};
+use Library;
+
+fn update_function_version(functions: &mut Vec<Function>, lib: *const Library) {
+    for function in functions {
+        let mut current_version = None;
+        for parameter in &function.parameters {
+            if let Some(new_version) = match (current_version,
+                                              match *unsafe { (*lib).type_(parameter.typ) } {
```
Comment (EPashkin) [inline src/update_version.rs]: IMHO this `lib` convert better do one time before `for`

Comment (EPashkin) [inline src/update_version.rs]: Or even do all hacks in `check_function_real_version` and pass normal reference here

**Comment by EPashkin** [comment]
Maybe checking return type also needed (Ex. g_test_dbus_new just luckily detect right version).
Don't see what other can be "improved"

**Comment by GuillaumeGomez** [comment]
Oh right! The return types! Completely forgot about them. Good catch!

File: src/update_version.rs (hunk 4 above)
Comment (GuillaumeGomez) [inline src/update_version.rs]: The result is a bit ugly by doing this way. I think it's preferable to keep it it as is.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
> I have worry how this works with crates with nonuniform versions,
> maybe this update need be configured and disabled by default.

Do you have an example of nonuniform versions?

**Comment by EPashkin** [comment]
We have https://github.com/gtk-rs/gdk-pixbuf/blob/master/Cargo.toml
that differs from other crates https://github.com/gtk-rs/gdk/blob/master/Cargo.toml#L27.
Cairo and Pango too.

**Comment by GuillaumeGomez** [comment]
We can add a "double-version-check = none" for some functions if you want?

**Comment by EPashkin** [comment]
IMHO better do it on crate level.

**Comment by EPashkin** [comment]
I mean user can disable all system if it starting detecting wrong.

**Comment by EPashkin** [comment]
Or we can limit checking to only main namespace

**Comment by GuillaumeGomez** [comment]
> Or we can limit checking to only main namespace

I think that's the better solution.

**Comment by GuillaumeGomez** [comment]
Limited to main namespace now.

File: src/update_version.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    if let Some(new_version) = new_version {
+        *current_version = Some(new_version);
+    }
+}
+
+fn update_function_version(functions: &mut Vec<Function>, lib: *const Library) {
+    for function in functions {
+        let mut current_version = None;
+        for parameter in &function.parameters {
+            check_versions(parameter, &mut current_version, lib);
```
Comment (EPashkin) [inline src/update_version.rs]: Before this better add check that `parameter.type.ns_id` is same as for function owner,
or nonuniform versions strikes ;)

Comment (GuillaumeGomez) [inline src/update_version.rs]: Oh right!

Comment (EPashkin) [inline src/update_version.rs]: Or better do it in `check_version`

Comment (GuillaumeGomez) [inline src/update_version.rs]: I did it in check_version. Better. ;)

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by EPashkin** [comment]
Good, now false gdk_pixbuf_loader_write_bytes limit to "v2_32" is gone.
It was wrong gdk as starting with 3.14 using gdk-pixbuf/v2_30, and no use next version.

**Comment by EPashkin** [comment]
@GuillaumeGomez Thanks.
Your prefer me merge as it or will do some squashing?

**Comment by GuillaumeGomez** [comment]
I have no problem with squash. What commits do you want to keep?

**Comment by EPashkin** [comment]
IMHO only last better separated.

**Comment by GuillaumeGomez** [comment]
Squashed! :)

**Comment by EPashkin** [comment]
Travis passed previous Appveyor too, so I merge this.

**Comment by GuillaumeGomez** [comment]
Yeay! Regen time! :D

**Comment by EPashkin** [comment]
Only `sys` actually need regen, maybe plus remove `version` for `Gio.Notification` in gio.

**Comment by sdroege** [comment]
Looks good to me too, even if a bit late :)

**Comment by EPashkin** [comment]
Late is better that never 😉

**Comment by GuillaumeGomez** [comment]
Wise @EPashkin has spoken!

**Pull request merged**